Physics

cNori opened the pull request:
# **Fixed joints including scale in calculations**
the PxJoint  is not expecting to get scaled offset and is collapsing auto anchor if A or B actor is scaled it will snap the joint in to wrong location

added SetTargetAnchor(const Transform& value)
# **Display Attached Colliders**
when selecting rigid body it will show all attached colliders for the rigidbody.
can be enabled or disabled per object using **DisplayAttachedColliders**
![image](https://github.com/FlaxEngine/FlaxEngine/assets/162097313/800cfba6-1891-49d6-9344-448459007030)
# **Remove thick frame from box collider**
when working with small enough collider it hard to set up good collisions for it
as a example 
![image](https://github.com/FlaxEngine/FlaxEngine/assets/162097313/da03383a-38e4-416f-bb40-df5eb4ac2f57)
this got me on side quest to re doing coliders _See:_ **Better colliders**
# **Better colliders**
Colliders
![image](https://github.com/FlaxEngine/FlaxEngine/assets/162097313/15f07604-a25d-4729-b89f-2bc22b088d46)
Triggers
![image](https://github.com/FlaxEngine/FlaxEngine/assets/162097313/e58116ba-8207-489c-8e31-de0a8dd1c61e)
- the colliders now can be displayed with out need to be selected with **Display Collider** 
- the style is not consistent across all colliders

**INFO:**
Debug shapes need some more work put in to look better (there is some over draw and weird shading)
# **Tasks list**
- [x] **Colliders no longer are needed to be directly under rb**
- [x] fixed incorent colliders when (A->B...->X) B.. is roteted
- [x] fix local center of mass is not calculated
- [x] (optional) fix collider passing through the ground with nested rigidbody when sub rb if IsKinematic is set to true

cNori committed:
fixed joints including scale in calculations

the PxJoint  is not expecting to get scaled offset and is collapsing auto anchor also if A nd B actor is scaled it will snap the joint in to wrong location

added SetTargetAnchor(const Transform& value)

cNori committed:
fixed compilation errors and changed one more spot

cNori committed:
Merge branch 'FlaxEngine:master' into Physics

cNori committed:
colliders will no longer need be directly under rigidbody

cNori committed:
AttachedTo editing protection hack

added a hack to prevent editing read only value from being edited

cNori committed:
Merge branch 'FlaxEngine:master' into Physics

cNori committed:
removed hack

cNori committed:
task done, fix incorent colliders when (A->B...->X) B.. is roteted

interesting case with colliders not being directly under rb, if the middle node('s) are rotated the collider was in wrong spot on back end

cNori committed:
fix "non-const lvalue reference to type..."

cNori committed:
Update Collider.cpp

cNori committed:
small code changes

cNori committed:
Better colliders,Display Attached Colliders

cNori committed:
Update RigidBody.cpp

cNori committed:
forgot about On Collider Changed ? in Collider::Attach(RigidBody* rigidBody)

cNori committed:
fixed center of mass offset, added seter for SetCenterOfMass

cNori committed:
sync

cNori committed:
fix missing overrides

warning: 'OnDebugDraw' overrides a member function but is not marked 'override'

cNori committed:
Merge branch 'FlaxEngine:master' into Physics

cNori committed:
Update PhysicsBackendPhysX.cpp

cNori committed:
Update RigidBody.cpp

cNori committed:
sync

cNori committed:
large worlds SetRigidDynamicActorCenterOfMass changed Float3 to Vector3

cNori committed:
Update PhysicsBackendEmpty.cpp

cNori committed:
added GetJointActorPose

cNori committed:
Create Deprecated Template Code.h

cNori committed:
Create PhysicsTransform.h

cNori committed:
sync

cNori committed:
Update AnimatedModelNode.cs

cNori committed:
fix compile errors

cNori committed:
update auto Anchor in editor

cNori committed:
fix compile errors

cNori committed:
reverted changes in Joint::OnTransformChanged()

cNori committed:
basic CylinderCollider

cNori committed:
Update Joint.h

cNori committed:
fix CS1041, C2065...

cNori committed:
Update Joint.h

cNori committed:
Update PxCylinderGeometry.cpp

cNori committed:
CylinderColl

cNori committed:
fix show A

cNori committed:
removed SetRigidActorPose from OnTransformChanged see note

cNori committed:
Update RigidBody.cpp

cNori committed:
sync

cNori committed:
Update Vector3.cs

cNori committed:
Update PhysicsTransform.cs

cNori committed:
Merge branch 'FlaxEngine:master' into Physics

cNori committed:
sync

cNori committed:
Merge remote-tracking branch 'upstream/master' into Physics

cNori commented:
[to do] 
the CylinderCollider is missing the GJK implementacion for scene querys
after a merge

Tryibion commented on Source/Engine/Physics/Actors/RigidBody.h:
```diff
@@ -39,6 +40,9 @@ class FLAXENGINE_API RigidBody : public Actor, public IPhysicsActor
 
+    friend Collider;
+    //API_FIELD(public,Attributes = "EditorOrder(100), DefaultValue(false), EditorDisplay(\"Rigid Body\"),RadyOnly")
+        Array<ScriptingObjectReference<Collider>> AttatchedColliders;
```
`AttachedColliders` instead of `AttatchedColliders`

Tryibion commented on Source/Engine/Physics/Actors/RigidBody.h:
```diff
@@ -472,17 +486,40 @@ class FLAXENGINE_API RigidBody : public Actor, public IPhysicsActor
 
     template<typename ColliderType = Collider, typename AllocationType = HeapAllocation>
-    void GetColliders(Array<ColliderType*, AllocationType>& result) const
+    void GetAttathedColliders(Array<ColliderType*, AllocationType>& result) const
```
AttachedColliders instead of AttatchedColliders

Tryibion commented on Source/Engine/Physics/Actors/RigidBody.h:
```diff
@@ -472,17 +486,40 @@ class FLAXENGINE_API RigidBody : public Actor, public IPhysicsActor
     }
 
+    template<typename ColliderType = Collider>
+    ColliderType* GetAttathedColliderOfType() const
```
AttachedColliders instead of AttathedColliders

Tryibion commented on Source/Engine/Physics/Colliders/Collider.cpp:
```diff
@@ -297,6 +313,23 @@ void Collider::OnMaterialChanged()
         PhysicsBackend::SetShapeMaterial(_shape, Material);
 }
 
+RigidBody* Collider::GetAttathmentRigidbody()
```
Attachment instead of Attathment

Tryibion commented on Source/Engine/Physics/Colliders/Collider.h:
```diff
@@ -156,11 +168,23 @@ API_CLASS(Abstract) class FLAXENGINE_API Collider : public PhysicsColliderActor
 
 private:
     void OnMaterialChanged();
+    RigidBody* GetAttathmentRigidbody();
```
Attachment instead of Attathment

Tryibion commented on Source/Engine/Physics/Joints/Joint.cpp:
```diff
@@ -16,10 +19,111 @@ Joint::Joint(const SpawnParams& params)
+        );
+    }
+}
+#pragma region SheredCode
+#define ImplementSheredCodeForActor(N)\
```
is this suppose to be Shared or sheared?

cNori replied:
o ye 😂 compiler will not care
but it can be changed

Tryibion commented on Source/Engine/Physics/Joints/Joint.cpp:
```diff
@@ -16,10 +19,111 @@ Joint::Joint(const SpawnParams& params)
+    return 0;\
+}
+#pragma endregion
+ImplementSheredCodeForActor(0)
+ImplementSheredCodeForActor(1)
+
```
is this Shared or sheared?

Tryibion commented on Source/Engine/Physics/Joints/Joint.cpp:
```diff
@@ -224,15 +281,90 @@ void Joint::DrawPhysicsDebug(RenderView& view)
-    DEBUG_DRAW_WIRE_SPHERE(BoundingSphere(GetTargetPosition(), 4.0f), Color::AliceBlue * 0.8f, 0, false);
+    auto p0 = GetPoseActor0();
+    auto p1 = GetPoseActor1();
+    if (Actor0) {
+        DEBUG_DRAW_WIRE_BOX(Actor0->GetBox(), Color::BlueViolet * 0.8f, 0, false);
+    }
```
Move starting bracket to new line for code consistency, or since it is only a single line of code after, remove brackets.

Tryibion commented on Source/Engine/Physics/Joints/Joint.h:
```diff
@@ -23,23 +34,140 @@ API_CLASS(Abstract) class FLAXENGINE_API Joint : public Actor
-    API_PROPERTY(Attributes="EditorOrder(10), DefaultValue(float.MaxValue), EditorDisplay(\"Joint\"), ValueCategory(Utils.ValueCategory.Force)")
-    FORCE_INLINE float GetBreakForce() const
+    API_PROPERTY(Attributes = "EditorOrder(10), DefaultValue(float.MaxValue), EditorDisplay(\"Joint\"), ValueCategory(Utils.ValueCategory.Force)")
+        FORCE_INLINE float GetBreakForce() const
```
No need to indent here

cNori replied:
auto formatted by vs

Tryibion commented on Source/Engine/Physics/Joints/Joint.h:
```diff
@@ -52,8 +180,8 @@ API_CLASS(Abstract) class FLAXENGINE_API Joint : public Actor
-    API_PROPERTY(Attributes="EditorOrder(20), DefaultValue(float.MaxValue), EditorDisplay(\"Joint\"), ValueCategory(Utils.ValueCategory.Torque)")
-    FORCE_INLINE float GetBreakTorque() const
+    API_PROPERTY(Attributes = "EditorOrder(20), DefaultValue(float.MaxValue), EditorDisplay(\"Joint\"), ValueCategory(Utils.ValueCategory.Torque)")
+        FORCE_INLINE float GetBreakTorque() const
```
No need to indent here

cNori replied:
auto formatted by vs

Tryibion commented on Source/Engine/Physics/Joints/Joint.h:
```diff
@@ -66,8 +194,8 @@ API_CLASS(Abstract) class FLAXENGINE_API Joint : public Actor
-    API_PROPERTY(Attributes="EditorOrder(30), DefaultValue(true), EditorDisplay(\"Joint\")")
-    FORCE_INLINE bool GetEnableCollision() const
+    API_PROPERTY(Attributes = "EditorOrder(30), DefaultValue(true), EditorDisplay(\"Joint\")")
+        FORCE_INLINE bool GetEnableCollision() const
```
No need to indent here

cNori replied:
auto formatted by vs

Tryibion commented on Source/Engine/Physics/Joints/Joint.h:
```diff
@@ -80,109 +208,99 @@ API_CLASS(Abstract) class FLAXENGINE_API Joint : public Actor
-    API_PROPERTY(Attributes="EditorOrder(39), DefaultValue(false), EditorDisplay(\"Joint\")")
-    bool GetEnableAutoAnchor() const;
+    API_PROPERTY(Attributes = "EditorOrder(39), DefaultValue(false), EditorDisplay(\"Joint\")")
+        bool GetEnableAutoAnchor() const;
```
No need to indent here

Tryibion commented on Source/Engine/Physics/Joints/Joint.h:
```diff
@@ -201,5 +319,57 @@ API_CLASS(Abstract) class FLAXENGINE_API Joint : public Actor
+    /// </summary>
+    API_FUNCTION() DEPRECATED void SetJointOrientation(const Quaternion& orientation);
+#else
+#ifndef STRING2
+#define STRING2(x) #x
+#define STRING(x) STRING2(x)
+#endif
+#pragma message ( __FILE__ "(" STRING(__LINE__) "):" "[Code Mantening] Remove Deprecated code")
+#endif
```
Is this needed?

cNori replied:
yes there is a code inside the engine which is dedicated and expired 
this is reminder

Tryibion commented on Source/Engine/Physics/PhysX/PhysicsBackendPhysX.cpp:
```diff
@@ -976,6 +984,39 @@ void PhysicalMaterial::UpdatePhysicsMaterial()
 
 bool CollisionCooking::CookConvexMesh(CookingInput& input, BytesContainer& output)
 {
+#if 0 // data extractor
+
+    String out = String::Format(L"\n static Float3* points [{0}] = ", input.VertexCount);
+    out.Append('{');
+    auto data = Array<Float3>(input.VertexData, input.VertexCount);
+    for (size_t i = 0; i < data.Count(); i++)
+    {
+        auto d = data[i];
+        out.Append(String::Format(L"\tFloat3({0},{1},{2}),\n", d.X, d.Y, d.Z));
+    }
+    out.Append(L"};");
+    if (input.Is16bitIndexData)
+    {
+        auto data2 = Array<PxU16>((PxU16*)input.IndexData, input.IndexCount);
+        for (size_t i = 0; i < data2.Count(); i++)
+        {
+            auto d = data2[i];
+            out.Append(String::Format(L"{0},\n",d));
+        }
+    }
+    else
+    {
+        auto data2 = Array<PxU32>((PxU32*)input.IndexData, input.IndexCount);
+        for (size_t i = 0; i < data2.Count(); i++)
+        {
+            auto d = data2[i];
+            out.Append(String::Format(L"{0},\n", d));
+        }
+    }
+
+    LOG_STR(Info, out.ToString());
+#endif
+
```
I assume this is not used? does it need to stay? maybe for debug purposes?

cNori replied:
> I assume this is not used?

yes it is unused

> does it need to stay?

idk i used it to extract verts for breaking and just left it there

> maybe for debug purposes?
if somfing break it will have use 
note is striped from compilation any way

Tryibion commented on Source/Templates/Deprecated Template Code.h:
```diff
@@ -0,0 +1,25 @@
+//template code for Deprecated code at some version
+//-----------------------------------------------------------------------------------
+//                         find and replace, legend / info
+//-----------------------------------------------------------------------------------
+// <depMAJOR> major version (0-...)
+// <depMINOR> minor version (0-9)
+//-----------------------------------------------------------------------------------
+//                                 requerd inclide
+//-----------------------------------------------------------------------------------
+// FlaxEngine.Gen.h
+//-----------------------------------------------------------------------------------
+
+
+#pragma region Deprecated
+#if (FLAXENGINE_VERSION_MAJOR != <depMAJOR> && (FLAXENGINE_VERSION_MINOR >= <depMINOR>))
+//<Deprecated code here>
+
+#else
+#ifndef STRING2
+#define STRING2(x) #x
+#define STRING(x) STRING2(x)
+#endif
+#pragma message ( __FILE__ "(" STRING(__LINE__) "):" "[Code Mantening] Remove Deprecated code")
+#endif
+#pragma endregion
```
Guess Mafi can decide if he wants this.

Tryibion reviewed (commented):
Some small spelling mistakes and a couple questions.

cNori committed:
rename

cNori committed:
fixing nitpicks

cNori committed:
changed the default of DisplayAttachedColliders to be false

cNori committed:
Merge branch 'FlaxEngine:master' into Physics

cNori committed:
Update Collider.cpp

cNori committed:
Update BehaviorTreeNodes.cpp

cNori committed:
Update PhysicsTransform.h

cNori committed:
expanded macros (IOS invalid preprocessing token)

cNori committed:
swaprd the imp of custom Cylinder to px version

throwed on backlog might be done some day

cNori committed:
cleen up

cNori committed:
sync

cNori committed:
Revert "rename"

This reverts commit 540dd4682110c332096426347bc62f4dbb5786f6.

cNori committed:
sync

fixed
joints location with out A or B actors set

CylinderCollider
now is using corent Margin from base class

rb kinematic
now is uspdated

cNori committed:
Update RigidBody.cpp

cNori committed:
PlaneCollider and ConeCollider

mafiesto4 commented:
Thanks for this PR! It's very big and it will take more time to review, merge and test code so I move it for future. Consider splitting it into smaller chunks.

cNori commented:
> Thanks for this PR! It's very big and it will take more time to review, merge and test code so I move it for future. Consider splitting it into smaller chunks.

this is close to the smallest chunk it can be

Tryibion commented:
I think some parts of this would be merged faster if it was broken up into 3 different PRs.
1. Changing the debug shapes/line size.
2. Fixes to rigid bodies and joints.
3. Additional physics colliders (Cylinder, Plane,  Cone, etc.).

cNori commented:
> I think some parts of this would be merged faster if it was broken up into 3 different PRs.
> 
> 1. Changing the debug shapes/line size to the new proposed style.
> 2. Fixes to rigid bodies and joints.
> 3. Additional physics colliders (Cylinder, Plane,  Cone, etc.).

some code is overlapping so idk about it

Tryibion committed:
Small style fixes to comments and Depricated code.

Tryibion committed:
Small code style fixes

Tryibion commented:
@mafiesto4, I tested this with the physics sample. Besides breaking the actor links in the joints, everything seems to work.

cNori commented:
> @mafiesto4, I tested this with the physics sample. Besides breaking the actor links in the joints, everything seems to work. I did not test the added collider shapes.

the actor joints might need to be re done but it not suposto break a lot unless there is a script is forcing some transform with is not compatibile with new system (scale is incuded in calculation for auto anchor)

cNori commented:
@Tryibion (directory after loading with new changes)
![image](https://github.com/user-attachments/assets/000fa9bc-8682-4c51-8dbd-1a6ed4ab7cce)

location of the joints position A is at 0,0,0
because of Actor 0 being set to none 
![image](https://github.com/user-attachments/assets/05a73666-4e16-4cb2-99fd-2dbaf198ec89)

allso the debug shows (need to update it some day so it is more use full)
![image](https://github.com/user-attachments/assets/8d79e39b-ba60-488f-8426-86f2b3ce42a2)
on some joints 

ragdroll is brokend for same reson Actor 0 being set to none 

some joints might be in wrong orietation do to change is in (scale and rotaction calculations)
it was broken so this is to be expected

cNori commented:
fixed and added the collider types and use features from the pr
![image](https://github.com/user-attachments/assets/b3cb88ec-736d-4c95-924f-048b648d649b)

[PhysicsFeaturesTour_Fixed.scene](https://github.com/user-attachments/files/18261648/Main.json)
ps. change extention to .scene

Tryibion commented:
See small changes from these review: <https://github.com/cNori/FlaxEngineFork/pull/1/commits/60cbc010f9e882d11d7ce394191abac8881f82ab> and <https://github.com/cNori/FlaxEngineFork/pull/1/commits/08861c33fb7b6069bd72f124ae62d12fc7f53f69>

cNori committed:
Merge branch 'master' into Physics

cNori committed:
fixed double def

cNori committed:
Merge branch 'master' into Physics

cNori committed:
Center of Mass Offset fix #3247

cNori committed:
Merge branch 'master' into Physics

cNori committed:
Update CharacterController.cpp

cNori commented:
i don't have time to maintain it and it was 2 yers at this point , @mafiesto4 pull what is needed from this pr have fun
